File: examples/ruby/benchmark.rb
require 'rubygems'
require 'bundler/setup'
require 'faye'

port   = ARGV[0] || 9292
path   = ARGV[1] || 'bayeux'
scheme = ARGV[2] == 'tls' ? 'https' : 'http'

EM.run {
  A = Faye::Client.new("#{scheme}://localhost:#{port}/#{path}")
  B = Faye::Client.new("#{scheme}://localhost:#{port}/#{path}")

  A.connect do
    B.connect do

      time = Time.now.to_f * 1000
      MAX  = 1000

      stop = lambda do
        puts Time.now.to_f * 1000 - time
        EM.stop
      end

      handle = lambda do |client, channel|
        lambda do |n|
          if n == MAX
            stop.call
          else
            client.publish(channel, n + 1)
          end
        end
      end

      sub_a = A.subscribe('/chat/a', &handle.call(A, '/chat/b'))
      sub_b = B.subscribe('/chat/b', &handle.call(B, '/chat/a'))

      sub_a.callback do
        sub_b.callback do
          puts 'START'
          A.publish('/chat/b', 0)
        end
      end
    end
  end
}

File: examples/node/benchmark.js
var faye = require('../../build/node/faye-node'),

    port   = process.argv[2] || 8000,
    path   = process.argv[3] || 'bayeux',
    scheme = process.argv[4] === 'tls' ? 'https' : 'http';

var A = new faye.Client(scheme + '://localhost:' + port + '/' + path),
    B = new faye.Client(scheme + '://localhost:' + port + '/' + path);

A.connect(function() {
  B.connect(function() {

    var time = new Date().getTime(),
        MAX  = 1000;

    var stop = function() {
      console.log(new Date().getTime() - time);
      process.exit();
    };

    var handle = function(client, channel) {
      return function(n) {
        if (n === MAX) return stop();
        client.publish(channel, n + 1);
      };
    };

    var subA = A.subscribe('/chat/a', handle(A, '/chat/b')),
        subB = B.subscribe('/chat/b', handle(B, '/chat/a'));

    subA.callback(function() {
      subB.callback(function() {
        console.log('START');
        A.publish('/chat/b', 0);
      });
    });
  });
});
